include/internal/refcount.h

```diff
@@ -193,14 +189,91 @@ static __inline int CRYPTO_GET_REF(volatile int *val, int *ret,
  */
 # ifndef HAVE_ATOMICS
 
-typedef int CRYPTO_REF_COUNT;
+typedef struct {
+    int val;
+#  ifdef OPENSSL_THREADS
+    CRYPTO_RWLOCK *lock;
+#  endif
+} CRYPTO_REF_COUNT;
+
+#  ifdef OPENSSL_THREADS
+
+static ossl_unused ossl_inline int CRYPTO_UP_REF(CRYPTO_REF_COUNT *refcnt, 
+                                                 int *ret)
+{
+    return CRYPTO_atomic_add(&refcnt->val, 1, ret, refcnt->lock);
+}
+
+static ossl_unused ossl_inline int CRYPTO_DOWN_REF(CRYPTO_REF_COUNT *refcnt,
+                                                   int *ret)
+{
+    return CRYPTO_atomic_add(&refcnt->val, -1, ret, refcnt->lock);
+}
+
+static ossl_unused ossl_inline int CRYPTO_GET_REF(CRYPTO_REF_COUNT *refcnt,
+                                                   int *ret)
+{
+    return CRYPTO_atomic_load_int(&refcnt->val, ret, refcnt->lock);
+}
+
+#   define CRYPTO_NEW_FREE_DEFINED  1
+static ossl_unused ossl_inline int CRYPTO_NEW_REF(CRYPTO_REF_COUNT *refcnt, int n)
```

Comment:
I think the n parameter should get dropped, for a reference count it should always be 1. All cases in the code where it's not 1 seems to be using a wrong API, and should use something like CRYPTO_atomic_add() instead.

Reply:
It is set to zero in `provider_core.c` for the activation count which seems like a legitimate use of a reference counter.  It could be changed to assign 1 and then decrement I suppose.

Reply:
As mentioned in some other PR, the activation count is not a reference count. It might look like one, but it's not. I have no idea if it's broken or not. If you want to be sure it works properly with relaxed semantics, this will need some careful audit and comments explaining why relaxed is ok and why it's so critical that it's relaxed. I prefer this to be handled with acquire+release semantics. It should be easy to write correct code with acquire+release semantics if you know what you're doing.

Note that x86(_64) does not actually do relaxed.

Reply:
Raised as #21279